pkg/modules/eventprocessing.go

```diff
@@ -0,0 +1,138 @@
+
+import (
+	"fmt"
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"runtime/debug"
+)
+
+// ApplyEventsSequentially takes a list of events and applies the given applyEvent function to each event in the list.
+// Processing is performed sequentially, one event at a time, in the order of the input list.
+// The EventLists returned by applyEvent are aggregated in a single EventList (in order of creation)
+// and returned by ApplyEventsSequentially.
+func ApplyEventsSequentially(
+	eventsIn *events.EventList,
+	applyEvent func(*eventpb.Event) (*events.EventList, error),
+) (*events.EventList, error) {
+
+	eventsOut := &events.EventList{}
+
+	iter := eventsIn.Iterator()
+	for event := iter.Next(); event != nil; event = iter.Next() {
+		evts, err := applyEvent(event)
+		if err != nil {
+			return nil, err
+		}
+		eventsOut.PushBackList(evts)
+	}
+
+	return eventsOut, nil
+}
+
+// ApplyEventsConcurrently takes a list of events and applies the given applyEvent function to each event in the list.
+// Processing is performed concurrently on all events in the input list.
+// Thus, the provided applyEvent function must be thread-safe.
+// The EventLists returned by applyEvent are aggregated in a single EventList and returned by ApplyEventsSequentially.
+// Despite being executed concurrently,
+// the order of the returned results preserves the order of the corresponding input events.
+// Thus, if applyEvent is deterministic, the output of ApplyEventsConcurrently is also deterministic.
+// If one or more errors occur during processing, ApplyEventsConcurrently returns the first of them,
+// along with an empty EventList.
+func ApplyEventsConcurrently(
+	eventsIn *events.EventList,
+	applyEvent func(*eventpb.Event) (*events.EventList, error),
+) (*events.EventList, error) {
+
+	// Common channels into which the results of each invocation of applyEvent will be written.
+	results := make(chan *events.EventList)
+	errors := make(chan error)
+
+	// For each input event, except for the last one, create a channel.
+	// This channel will be closed when the results of processing the event
+	// have been written to the common output channels.
+	// The i-th (except for the first) event-processing goroutine
+	// waits until the (i-1)-th one (in order of the input events)
+	// has written its output to the common channels before writing its own output.
+	// This ensures that the event processing results are sorted according to the corresponding input events.
+	done := make([]chan struct{}, eventsIn.Len()-1)
+	for i := 0; i < eventsIn.Len()-1; i++ {
+		done[i] = make(chan struct{})
+	}
+
+	// Start one concurrent worker for each event in the input list.
+	// Note that the processing starts concurrently for all events, and only the writing of the results is synchronized.
+	iter := eventsIn.Iterator()
+	i := 0
+	for event := iter.Next(); event != nil; event = iter.Next() {
+
+		go func(e *eventpb.Event, j int) {
```

Comment:
I suspect that for optimal performance it may be best to have a bounded number of worker goroutines, but I am not 100% sure and it would probably make the code a bit more complicated.

Reply:
Yes, I was thinking about that too, but for now I think what we have is good enough. From experience, the performance difference was marginal / not observable in an analogous case in a previous project. If this becomes an issue, we can easily limit the number of threads later.